### ISSUE TITLE
`c_unwind` full stabilization request: change in `extern "C"` behavior

### ISSUE THREAD

**Issue by BatmanAoD** [rust-lang/rust#115285]
Tracking issue: #74990

This is a separate issue to enable using the FCP bot a second time. The request-for-stabilization comment is duplicated below, without modification.

RFC "C-unwind ABI": https://github.com/rust-lang/rfcs/pull/2945

Feature gate to be stabilized: `#![feature(c_unwind)]`

**Comment by BatmanAoD** [comment]
# Request for Stabilization

This is a request for *full* stabilization of the `c_unwind` feature, [RFC-2945][rfc-text]. The behavior of non-`"Rust"`, non-`unwind` ABIs (such as `extern "C"`) will be modified to close soundness holes caused by permitting stack-unwinding to cross FFI boundaries that do not support unwinding.

## Summary

When using `panic=unwind`, if a Rust function marked `extern "C"` panics (and that panic is not caught), the runtime will now abort.

Previously, the runtime would simply attempt to unwind the caller's stack, but the behavior when doing so was undefined, because `extern "C"` functions are optimized with the assumption that they cannot unwind (i.e. in `rustc`, they are given the LLVM `nounwind` annotation).

This affects existing programs. If a program relies on a Rust panic "escaping" from `extern "C"`:
* It is currently unsound.
* Once this feature is stabilized, the program will crash when this occurs, whereas previously it may have appeared to work as expected.
* Replacing `extern "x"` with `extern "x-unwind"` will produce the intended behavior without the unsoundness.

The behavior of function calls using `extern "C"` is unchanged; thus, it is still undefined behavior to call a C++ function that throws an exception using the `extern "C"` ABI, even when compiling with `panic=unwind`.

## Changes since the RFC

None.

### Unresolved questions

This feature does not affect the behavior of `pthread_exit` or `longjmp`, so stabilization will neither break existing code using these features nor provide any soundness guarantee.

## Tests

### `"C"` guaranteed to abort on panic

[This test][abort-on-panic] verifies that an `extern "C"` function that `panic!`s will always abort if the panic would otherwise "escape".

## Documentation

The documentation PRs for this feature did not distinguish between the initial partially-stabilized behavior and the fully-stabilized behavior, so they already include the modifications to non-`unwind` ABIs.

* [The Rustonomicon PR][nomicon] has already been merged.
* [The Reference PR][reference] is still awaiting approval.

I've also [opened an issue][book-issue] to discuss adding a brief note about this feature in the Book.


[rfc-text]: https://github.com/rust-lang/rfcs/blob/master/text/2945-c-unwind-abi.md
[abort-on-panic]: https://github.com/rust-lang/rust/blob/master/tests/ui/panics/abort-on-panic.rs
[nomicon]: https://github.com/rust-lang/nomicon/pull/365
[reference]: https://github.com/rust-lang/reference/pull/1226
[book-issue]: https://github.com/rust-lang/book/issues/3729

**Comment by joshtriplett** [comment]
@rfcbot merge

**Comment by nbdd0121** [comment]
cc @rust-lang/wg-ffi-unwind

**Comment by tmandry** [comment]
> * Once this feature is stabilized, the program will crash when this occurs, whereas previously it may have appeared to work as expected.

Do we have a crater run showing the impact of this? I'd prefer to give the ecosystem time to update their code ahead of crashing everyone's programs (even if they're technically already broken).

That said, we will need some way of alerting all the broken libraries.

**Comment by nbdd0121** [comment]
We could do the crater run on the PR (still under preparation, sorry!)

**Comment by BatmanAoD** [comment]
I do think it makes sense to do a crater run before actually approving the stabilization.

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

Nominating so we can discuss what needs to happen to move this forward.

**Comment by nikomatsakis** [comment]
I'm in favor of doing this. At worst we can tie it to the edition.

**Comment by traviscross** [comment]
The situation seems to be that we're waiting on #116088 for the crater run, and that PR is waiting on #113923 to land.

**Comment by BatmanAoD** [comment]
@traviscross That's correct; sorry for not linking to those myself.

**Comment by nbdd0121** [comment]
Crater run is completed. I have [triaged](https://github.com/rust-lang/rust/pull/116088#issuecomment-1870577466) all crater failures and determined they are either flaky test, or broken crate, or are expected failure due to the change to enforce extern "C" to be nounwind.

**Comment by traviscross** [comment]
@rustbot labels +I-lang-nominated

Nominating for discussion as it seems we had been waiting on a crater run and analysis that now has been [completed](https://github.com/rust-lang/rust/pull/116088#issuecomment-1870577466).

**Comment by RalfJung** [comment]
However, https://github.com/rust-lang/rust/pull/113923 has also been [reverted](https://github.com/rust-lang/rust/pull/119885), so currently we can't stabilize this yet AFAIK.

**Comment by nbdd0121** [comment]
I think that's should only be blocking the PR, not the FCP?

**Comment by RalfJung** [comment]
> I'm in favor of doing this. At worst we can tie it to the edition.

@nikomatsakis so... would you like to check your box for FCP to start? :)

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

SO EXCITED

**Comment by BatmanAoD** [comment]
> expected failure due to the change to enforce extern "C" to be nounwind

I think this is what we were specifically worried about, isn't it? How widespread is this? (Sorry, I haven't looked at the data myself yet)

**Comment by cratelyn** [comment]
the `zcash` crate looks to be https://crates.io/crates/zcash a namesquatting crate. it is version 0.0.0, hasn't been updated in 5 years, and the associated repository has since been deleted.

p.s. _so_ excited to see this work in the final comment period! thank you all again for helping mentor my work on #76570. :heart_decoration:

**Comment by bjorn3** [comment]
The crater report contains regressions for several zcash related git repos: https://crater-reports.s3.amazonaws.com/pr-116088-2/index.html Several at least are forks of the upstream zcash, for which no regressions are reported. So it may well be that they were forked before it was fixed in upstream.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated

We discussed this in triage today with the result that this in now entering FCP.  Thanks to @BatmanAoD for working to push this forward.

**Comment by CAD97** [comment]
Small thing: I find it somewhat surprising the exact manner in which unwinds become aborts in `extern "C" fn`. Specifically, the current behavior is that each unwinding edge in the `extern "C" fn` is set to `terminate(abi)`, resulting in that any locals of the `extern "C" fn` itself are not dropped. The behavior that I expected is achievable by immediately delegating to a function with an unwinding ABI — that unwinding continues as normal up until it reaches the `extern "C"` boundary, at which point it aborts. (In MIR terms, generate cleanup bbs as normal, only replacing `unwind continue` with `unwind terminate(abi)`, leaving any `unwind: bbN` as-is.)

Further refining, it surprises me that wrapping a function body in an immediately invoked closure changes semantics. I do not expect the changed ABI in the signature to change the semantics of the contained body.

I could live with this, as the developer actually technically has more control in this setup, and it's lower overhead to abort more eagerly, but it still feels a bit wrong, and performance of the about-to-abort edge isn't particularly important. Any optimization barriers to the happy path are mostly still present due to the unwind edge existing at all.

**Comment by RalfJung** [comment]
I agree that's surprising, and opened an issue: https://github.com/rust-lang/rust/issues/123231.

Not sure if that's something we need to resolve before stabilization. However if we want to keep the current behavior we definitely need to document somewhere that calling these destructors is not guaranteed. I agree that's very surprising.

**Comment by daira** [comment]
> The crater report contains regressions for several zcash related git repos: https://crater-reports.s3.amazonaws.com/pr-116088-2/index.html Several at least are forks of the upstream zcash, for which no regressions are reported. So it may well be that they were forked before it was fixed in upstream.

Speaking as ECC's engineering manager responsible for zcashd and librustzcash, I *strongly* support stabilization of this change. In fact I have been [bitterly complaining](https://github.com/rust-lang/rust/issues/28728#issuecomment-842198878), with this issue as one of the examples, about the time it takes to fix soundness bugs in Rust. So it would be rather ironic if Zcash-related code were any part of the reason you were not doing so in this case (although I cannot see why it would be).

zcashd [pins a specific Rust version](https://github.com/zcash/zcash/blob/master/depends/packages/native_rust.mk), so there is no significant risk. If it breaks us, we will fix it before changing that pin &mdash; or not, because zcashd is being deprecated anyway.

We also have an [issue to change all our uses of "C" to "C-unwind"](https://github.com/zcash/librustzcash/issues/355) in librustzcash; the only reason why that was not prioritized is that there appeared to be no significant progress on the full stabilization, which is precisely the thing that would make "C-unwind" useful to us.

**Comment by BatmanAoD** [comment]
@daira

Believe me, I sympathize with the pain of how long this feature/fix has taken. 

> We also have an issue to change all our uses of "C" to "C-unwind" in librustzcash; the only reason why that was not prioritized is that there appeared to be no significant progress on the full stabilization, which is precisely the thing that would make "C-unwind" useful to us.

Sorry, I don't understand this dilemma; the "full stabilization" will only affect `"C"`. The `"C-unwind"` behavior, as described in the stabilization request and RFC, is already in place.
